### PR TITLE
WebM Video Builder

### PR DESCRIPTION
**Pull request by akre54**
# Add a VideoBuilder class

The web runs primarily realtime and interactive but there are a number of cases where you might want the content in the canvas to run out of realtime to render to a video. A very fast or very slow draw call, e.g.

This builds on the work of the GifBuilder class and adds a bit more info to the encoding-rfc. Please let me know if I should make any edits.

cc @chrisgervang and https://github.com/visgl/hubble.gl/issues/236

## Open questions

* Is it right to wrap the various APIs or should the client be responsible for passing in a buffer? I'm leaning towards ease-of-use for common cases, but the client can also pass in a buffer.

* Should this use the frameBuffer instead of using readPixels? Should it use readPixelsToBuffer?

## Alternatives considered

* MediaRecorder + canvas.getStream - only realtime
* WebMWriter - uses canvas.toDataUrl() which incurrs costly serialization
* VideoFrame - not supported in FireFox

### CODE REVIEW

File: modules/video/src/video-builder.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    } else if (source instanceof HTMLCanvasElement) {
+      buffer = source.getContext('2d')!.getImageData(0, 0, source.width, source.height).data;
+    } else if (source instanceof CanvasRenderingContext2D) {
+      buffer = source.getImageData(0, 0, source.canvas.width, source.canvas.height).data;
+    } else if (
+      source instanceof WebGLRenderingContext ||
+      source instanceof WebGL2RenderingContext
+    ) {
+      const gl = source;
+      // const {width, height} = gl.canvas;
```
Comment (akre54) [inline modules/video/src/video-builder.ts:56]: Either one is fine with me. Preference?

Comment (chrisgervang) [reply modules/video/src/video-builder.ts:56]: Either should be fine if they're both the buffer size. I've assumed at this point `const {width, height} = gl.canvas;` was the way to do it, but I only have limited theoretical knowledge of how it differs from `GL.VIEWPORT`. 

Do you know if `GL.VIEWPORT` is meaningfully different? The [spec](https://www.khronos.org/registry/webgl/specs/latest/1.0/#2.3) suggests they are synced on the application level.

Pulling the correct size has been a source of issues in the past due to subtle differences in how libs handle it. Luma.gl uses [`gl.canvas`](https://github.com/visgl/luma.gl/blob/c3e42ceb635ef0a698f742f27a0c059b3a0244c0/modules/webgl/src/context/context/device-pixels.ts#L80-L81). Mapbox did something different. Maplibrae has diverged now I think. I tried to sort it out in this [issue](https://github.com/visgl/hubble.gl/issues/159) and [rfc](https://github.com/visgl/hubble.gl/pull/176/files), but I wouldn't be surprised if it comes up again in some corner case.

I say go with what you know. I only know enough to flag it might not be straight forward.

File: modules/video/test/video-builder.spec.js [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      videoBuilder.addFrame(gl);
+
+      if (tick > 120) {
+        loop.delete();
+      }
+    },
+
+    async onFinalize({gl}) {
+      const videoDataUrl = await videoBuilder.finalize();
+      t.ok(videoDataUrl, 'finalize() returns WebM video URL');
```
[the remark this answers is not in the document]
Comment (akre54) [reply modules/video/test/video-builder.spec.js:110]: Yeah my thought here was to use the SnapshotTestRunner and then visually compare a couple of the frames (say, first, middle, last?). What do you think?

**Comment by chrisgervang** [comment]
Regarding the open questions,

- Is it right to wrap the various APIs or should the client be responsible for passing in a buffer? I'm leaning towards ease-of-use for common cases, but the client can also pass in a buffer.
  - I think this is fine since a buffer is one of the supported options, and this removes complexity for the client.
- Should this use the frameBuffer instead of using readPixels? Should it use readPixelsToBuffer?
  - Is [`readPixelsToBuffer`](https://github.com/visgl/luma.gl/blob/729878e3a8f33981287e5a788a143abf94c265f6/modules/gltools/src/classic/copy-and-blit.ts#L87) from luma? I had to google it.
  - I would think handling a custom frameBuffer input is another branch of logic.
  - Is `readPixels` better than `canvas.toBlob` since it's more direct?

File: modules/video/test/video-builder.spec.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrisgervang) [reply modules/video/test/video-builder.spec.js:110]: That sounds solid. Very curious if compression is deterministic or flaky.

File: modules/video/docs/api-reference/video-builder.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
+# VideoBuilder
```
Comment (ibgreen) [inline modules/video/docs/api-reference/video-builder.md:1]: I suspect this need to be added to `docs/table-of-contents.json`.

File: modules/video/docs/api-reference/video-builder.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+<p class="badges">
+  <img src="https://img.shields.io/badge/From-v3.2-blue.svg?style=flat-square" alt="From-v3.2" />
+</p>
+
+> The `VideoBuilder` is experimental.
```
Comment (ibgreen) [inline modules/video/docs/api-reference/video-builder.md:7]: Should it be called `WebMBuilder` or `WebMVideoBuilder`?

File: modules/video/src/video-builder.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+type VideoBuilderOptions = Partial<typeof VIDEO_BUILDER_OPTIONS>;
+
+function nextEvent(target, name): Promise<{data: any}> {
+  return new Promise((resolve) => {
+    target.addEventListener(name, resolve, {once: true});
+  });
+}
+
+export default class VideoBuilder {
+  worker: Worker = new Worker('node_modules/webm-wasm/dist/webm-worker.js');
```
Comment (ibgreen) [inline modules/video/src/video-builder.ts:18]: We typically include such workers in our own dist, rather than import them from installed packages. Given that it is apache licensed there are no concerns with copying it, especially if we give some attribution.

The `libs` folder in each module is the place to put such "binaries". There are a bunch of modules that do this.


The downside is we don't automatically get access to patches but I think it may cause problems for users if we depend on an npm install tree.

File: modules/video/docs/api-reference/video-builder.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+## Options
+
+| Option | Type | Default | Description |
+| ------ | ---- | ------- | ----------- |
+| `width` | number | 512 | The height of the output video
+| `height` | number | 512 | The width of the output video
+| `framerate` | number | 24 | How many frames per second the final container should hold
+| `bitrate` | number | 200 | The bitrate (in kbps) to record the
+| `realtime` | boolean | false | Whether to record in realtime. Used for streaming videos, for instance.
```
Comment (ibgreen) [inline modules/video/docs/api-reference/video-builder.md:43]: Add an attribution at the bottom perhaps if we include the wasm see below.

File: modules/video/src/video-builder.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+      id: 'webm',
+      name: 'WEBM',
+      extensions: ['webm'],
+      mimeTypes: ['video/webm'],
+      builder: VideoBuilder,
+      options: VIDEO_BUILDER_OPTIONS
+    };
+  }
+
+  async initialize(options: VideoBuilderOptions): Promise<void> {
```
Comment (ibgreen) [inline modules/video/src/video-builder.ts:31]: Maybe a JSDoc `/** */` comment on each method, it is nice when hovering in vscode if nothing else...

File: modules/video/src/video-builder.ts (hunk 1 above)
Comment (ibgreen) [reply modules/video/src/video-builder.ts:56]: 1. loaders.gl is generally WebGL agnostic - it doesn't depend on WebGL but works with typed arrays which are optimized for WebGL use.
2. Also loaders.gl doesn't use Node.js constructs like `Buffer`, preferring to use typed arrays in all interfaces. Simply mentioning `Buffer` tends to pull in a 50KB polyfill in browser bundles...

Comment (ibgreen) [reply modules/video/src/video-builder.ts:56]: I would have liked it more if this function just accepted a typed array and if necessary we offered some utility function for getting hold of a typed array from different input types.

**Review by ibgreen — APPROVED** [review]
Nice addition! Sorry for delay, I am sure I reviewed this a few days ago but somehow my comments didn't make it in.

File: modules/video/src/video-builder.ts (hunk 5 above)
Comment (akre54) [reply modules/video/src/video-builder.ts]: Done. It seems like they don't update all that regularly, so we should be fine for the time being. 

That said, I'm having trouble with the webpack config here... doesn't seem to be finding the worker code when I try to import (I see there's a rule for `test: /\.worker\.js$/,` but it doesn't seem to be working). From what I can tell from the parquet-wasm and draco loader, it looks like they're loaded from third party CDNs. Any pointers on what to do here?

File: modules/video/docs/api-reference/video-builder.md (hunk 6 above)
Comment (akre54) [reply modules/video/docs/api-reference/video-builder.md]: Good call. Done.

File: modules/video/src/video-builder.ts (hunk 7 above)
Comment (akre54) [reply modules/video/src/video-builder.ts]: Sounds good. Will add.

**Review by akre54 — COMMENTED** [review]
Updated with feedback. Worker URL situation is still a little wonky, would love some help or pointers here.

File: modules/video/src/webm-video-builder.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
+const VIDEO_BUILDER_OPTIONS = {
+  width: 512,
+  height: 512,
+  framerate: 24,
+  bitrate: 200,
+  realtime: false
+};
+
+type VideoBuilderOptions = Partial<typeof VIDEO_BUILDER_OPTIONS>;
```
Comment (chrisgervang) [inline modules/video/src/webm-video-builder.ts:9]: `WebMVideoBuilderOptions`

File: modules/video/src/webm-video-builder.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
+const VIDEO_BUILDER_OPTIONS = {
```
Comment (chrisgervang) [inline modules/video/src/webm-video-builder.ts:1]: `WEBM_VIDEO_BUILDER_OPTIONS`

**Review by chrisgervang — APPROVED** [review]

**Comment by akre54** [comment]
Any thoughts about loading the worker file? Right now it depends on a specific hard coded path exposed through the server.

I couldn't get the Webpack loader to require the file (I gave up debugging and didn't look too deeply into the config).

**Comment by chrisgervang** [comment]
> Any thoughts about loading the worker file? Right now it depends on a specific hard coded path exposed through the server.
> 
> I couldn't get the Webpack loader to require the file (I gave up debugging and didn't look too deeply into the config).

How does the rest of loaders handle this, @ibgreen? It's not clear to me if this will this cause any issues as-is for users?

**Comment by ibgreen** [comment]
The use case is still relevant, but this implementation is not anymore.

The browser media stack has evolved substantially since this was opened in 2022:

- WebCodecs now provides `VideoFrame` and `VideoEncoder`, including explicit timestamps suitable for faster- or slower-than-real-time rendering.
- [Mediabunny's `CanvasSource`](https://mediabunny.dev/guide/media-sources#canvassource) already covers the higher-level workflow here: add timestamped canvas frames, encode them, and mux MP4 or WebM to memory or a stream.
- [luma.gl's current video-texture example](https://github.com/visgl/luma.gl/blob/master/examples/api/video-texture/app.ts) is already using Mediabunny, `VideoFrame`, `captureStream(0)`, and `requestFrame()`.

This patch, meanwhile, vendors the old `webm-wasm` proof of concept, remains WebM-specific, and still has an unresolved hard-coded worker URL.

If loaders.gl revisits this feature, it should be a generic `VideoWriter` based on WebCodecs plus a modern muxer—or simply documentation showing direct Mediabunny integration—rather than reviving this implementation.

Closing as superseded. Thanks for the original contribution and discussion!

**Comment by akre54** [comment]
Thanks Ib. Agreed this doesn't make sense for Luma now (and might not have before either). Fwiw I'm using exactly that stack for Noodles.gl! These days I usually just tell the AI to basically clone Noodles
